File: Tools/CodeMetrics/PageValidator/PyPageValidator/Tokenizer/tokenizer.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Optional, Tuple


# ---------------------------------------------------------------------------
# Token hierarchy — one subclass per lexical unit.
# ---------------------------------------------------------------------------

class Token:
    pass

@dataclass
class TagOpen(Token):
    name: str

@dataclass
class TagClose(Token):
    name: str

@dataclass
class AttrName(Token):
    name: str

@dataclass
class AttrValue(Token):
    value: str          # quoted

@dataclass
class AttrUnquoted(Token):
    value: str          # unquoted

class SelfClose(Token):
    pass

class TagEnd(Token):
    pass

@dataclass
class Text(Token):
    content: str

@dataclass
class Comment(Token):
    content: str

@dataclass
class Doctype(Token):
    content: str

class Eof(Token):
    pass


# ---------------------------------------------------------------------------
# Tokenizer
# ---------------------------------------------------------------------------

class Tokenizer:
    def __init__(self, src: str) -> None:
        self._src = src
        self._pos = 0
        self._line = 1
        self._col = 1
        self._in_tag = False
        self._pending: Optional[Tuple[Token, Tuple[int, int]]] = None
        self._last_start: Tuple[int, int] = (1, 1)

    @property
    def token_start(self) -> Tuple[int, int]:
        return self._last_start

    def next_token(self) -> Token:
        if self._pending is not None:
            tok, pos = self._pending
            self._pending = None
            self._last_start = pos
            return tok

        self._last_start = (self._line, self._col)

        if self._in_tag:
            return self._scan_in_tag()

        if self._pos >= len(self._src):
            return Eof()

        if self._src[self._pos] == '<':
            self._advance()
            if self._pos < len(self._src) and self._src[self._pos] == '!':
                self._advance()
                if (self._pos + 1 < len(self._src) and
                        self._src[self._pos] == '-' and
                        self._src[self._pos + 1] == '-'):
                    self._advance()
                    self._advance()
                    return Comment(self._collect_until('-->'))
                return Doctype(self._collect_until('>'))
            if self._pos < len(self._src) and self._src[self._pos] == '/':
                self._advance()
                name = self._collect_name()
                self._skip_ws()
                if self._pos < len(self._src) and self._src[self._pos] == '>':
                    self._advance()
                return TagClose(name)
            tag_name = self._collect_name()
            self._in_tag = True
            return TagOpen(tag_name)

        buf: list[str] = []
        while self._pos < len(self._src) and self._src[self._pos] != '<':
            buf.append(self._advance())
        return Text(''.join(buf))

    # -----------------------------------------------------------------------

    def _peek(self, off: int = 0) -> str:
        i = self._pos + off
        return self._src[i] if i < len(self._src) else '\0'

    def _advance(self) -> str:
        if self._pos >= len(self._src):
            return '\0'
        c = self._src[self._pos]
        self._pos += 1
        if c == '\n':
            self._line += 1
            self._col = 1
        else:
            self._col += 1
        return c

    def _skip_ws(self) -> None:
        while self._pos < len(self._src) and self._src[self._pos].isspace():
            self._advance()

    def _collect_name(self) -> str:
        buf: list[str] = []
        while self._pos < len(self._src):
            c = self._src[self._pos]
            if c.isalnum() or c in '-_:.':
                buf.append(self._advance())
            else:
                break
        return ''.join(buf)

    def _collect_until(self, stop: str) -> str:
        buf: list[str] = []
        n = len(stop)
        while self._pos < len(self._src):
            if self._src[self._pos:self._pos + n] == stop:
                for _ in stop:
                    self._advance()
                break
            buf.append(self._advance())
        return ''.join(buf)

    def _scan_in_tag(self) -> Token:
        self._skip_ws()
        if self._pos >= len(self._src) or self._src[self._pos] == '>':
            if self._pos < len(self._src):
                self._advance()
            self._in_tag = False
            return TagEnd()
        if self._src[self._pos] == '/' and self._peek(1) == '>':
            self._advance()
            self._advance()
            self._in_tag = False
            return SelfClose()
        name = self._collect_name()
        if not name:
            self._advance()
            return self._scan_in_tag()
        self._skip_ws()
        if self._peek() != '=':
            return AttrName(name)
        self._advance()  # '='
        self._skip_ws()
        val_pos = (self._line, self._col)
        q = self._peek()
        if q in ('"', "'"):
            self._advance()  # opening quote
            val: list[str] = []
            while self._pos < len(self._src) and self._src[self._pos] != q:
                val.append(self._advance())
            if self._pos < len(self._src):
                self._advance()  # closing quote
            self._pending = (AttrValue(''.join(val)), val_pos)
        else:
            val = []
            while self._pos < len(self._src):
                c = self._src[self._pos]
                if c.isspace() or c in '>/':
                    break
                val.append(self._advance())
            self._pending = (AttrUnquoted(''.join(val)), val_pos)
        return AttrName(name)

File: Tools/CodeMetrics/PageValidator/CsPageValidator/Tokenizer/Tokenizer.cs
namespace PageValidator;

// ---------------------------------------------------------------------------
// Token discriminated union — one subtype per lexical unit.
// ---------------------------------------------------------------------------

public abstract record Token;

public sealed record TagOpen(string Name)    : Token;
public sealed record TagClose(string Name)   : Token;
public sealed record AttrName(string Name)   : Token;
public sealed record AttrValue(string Value) : Token;      // quoted
public sealed record AttrUnquoted(string Value) : Token;   // unquoted
public sealed record SelfClose               : Token;
public sealed record TagEnd                  : Token;
public sealed record Text(string Content)    : Token;
public sealed record Comment(string Content) : Token;
public sealed record Doctype(string Content) : Token;
public sealed record Eof                     : Token;

// ---------------------------------------------------------------------------
// Tokenizer
// ---------------------------------------------------------------------------

public sealed class Tokenizer
{
    private readonly string _src;
    private int _pos;
    private int _line;
    private int _col;
    private bool _inTag;
    private (Token Tok, (int Line, int Col) Pos)? _pending;
    private (int Line, int Col) _lastStart;

    public Tokenizer(string src)
    {
        _src       = src;
        _pos       = 0;
        _line      = 1;
        _col       = 1;
        _inTag     = false;
        _lastStart = (1, 1);
    }

    public (int Line, int Col) TokenStart => _lastStart;

    public Token NextToken()
    {
        if (_pending.HasValue)
        {
            var (tok, pos) = _pending.Value;
            _pending = null;
            _lastStart = pos;
            return tok;
        }
        _lastStart = (_line, _col);
        if (_inTag) return ScanInTag();
        if (_pos >= _src.Length) return new Eof();

        if (_src[_pos] == '<')
        {
            Advance();
            if (_pos < _src.Length && _src[_pos] == '!')
            {
                Advance();
                if (_pos + 1 < _src.Length && _src[_pos] == '-' && _src[_pos + 1] == '-')
                {
                    Advance(); Advance();
                    return new Comment(CollectUntil("-->"));
                }
                return new Doctype(CollectUntil(">"));
            }
            if (_pos < _src.Length && _src[_pos] == '/')
            {
                Advance();
                var name = CollectName();
                SkipWs();
                if (_pos < _src.Length && _src[_pos] == '>') Advance();
                return new TagClose(name);
            }
            var tagName = CollectName();
            _inTag = true;
            return new TagOpen(tagName);
        }

        var sb = new System.Text.StringBuilder();
        while (_pos < _src.Length && _src[_pos] != '<')
            sb.Append(Advance());
        return new Text(sb.ToString());
    }

    // -----------------------------------------------------------------------

    private char Peek(int off = 0)
    {
        int i = _pos + off;
        return i < _src.Length ? _src[i] : '\0';
    }

    private char Advance()
    {
        if (_pos >= _src.Length) return '\0';
        char c = _src[_pos++];
        if (c == '\n') { _line++; _col = 1; }
        else            { _col++; }
        return c;
    }

    private void SkipWs()
    {
        while (_pos < _src.Length && char.IsWhiteSpace(_src[_pos]))
            Advance();
    }

    private string CollectName()
    {
        var sb = new System.Text.StringBuilder();
        while (_pos < _src.Length)
        {
            char c = _src[_pos];
            if (char.IsLetterOrDigit(c) || c == '-' || c == '_' || c == ':' || c == '.')
                sb.Append(Advance());
            else
                break;
        }
        return sb.ToString();
    }

    private string CollectUntil(string stop)
    {
        var sb = new System.Text.StringBuilder();
        while (_pos < _src.Length)
        {
            if (_pos + stop.Length <= _src.Length &&
                _src.AsSpan(_pos, stop.Length).SequenceEqual(stop.AsSpan()))
            {
                for (int i = 0; i < stop.Length; i++) Advance();
                break;
            }
            sb.Append(Advance());
        }
        return sb.ToString();
    }

    private Token ScanInTag()
    {
        SkipWs();
        if (_pos >= _src.Length || _src[_pos] == '>')
        {
            if (_pos < _src.Length) Advance();
            _inTag = false;
            return new TagEnd();
        }
        if (_src[_pos] == '/' && Peek(1) == '>')
        {
            Advance(); Advance();
            _inTag = false;
            return new SelfClose();
        }
        var name = CollectName();
        if (name.Length == 0)
        {
            Advance();
            return ScanInTag();
        }
        SkipWs();
        if (Peek() != '=')
            return new AttrName(name);
        Advance(); // '='
        SkipWs();
        var valPos = (_line, _col);
        char q = Peek();
        if (q == '"' || q == '\'')
        {
            Advance(); // opening quote
            var val = new System.Text.StringBuilder();
            while (_pos < _src.Length && _src[_pos] != q)
                val.Append(Advance());
            if (_pos < _src.Length) Advance(); // closing quote
            _pending = (new AttrValue(val.ToString()), valPos);
        }
        else
        {
            var val = new System.Text.StringBuilder();
            while (_pos < _src.Length)
            {
                char c = _src[_pos];
                if (char.IsWhiteSpace(c) || c == '>' || c == '/') break;
                val.Append(Advance());
            }
            _pending = (new AttrUnquoted(val.ToString()), valPos);
        }
        return new AttrName(name);
    }
}
